### PR TITLE
IBX-5022 - Checkout customization

### PR DESCRIPTION
**Pull request by dabrt**
| Question      | Answer |
| ------------- | --- |
| JIRA Ticket   | [IBX-5022](https://issues.ibexa.co/browse/IBX-5022) |
| Versions      |4.4 and up |
| Edition       | Commerce |

Add article on checkout customization

#### Checklist

- [x] Text renders correctly
- [x] Text has been checked with vale
- [x] Description metadata is up to date
- [x] Code samples are working
- [ ] PHP code samples have been fixed with PHP CS fixer
- [x] Added link to this PR in relevant JIRA ticket or code PR

### CODE REVIEW

**Review by GrabowskiM — APPROVED** [review]

File: docs/commerce/checkout/customize_checkout.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+For example, if the website is selling airline tickets, you may need several additional steps 
+with passengers defining their special needs.
+On the other side of the spectrum would be a store that sells books with personal pickup, 
+where one-page checkout would be enough.
+
+There are several factors that make checkout particularly flexible and customizable:
+
+- it is based on Symfony workflow
+- exposes a variety of APIs
+- exposes Twig functions that help you render the steps
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:30]:
```suggestion
- it exposes a variety of APIs
- it exposes Twig functions that help you render the steps
```

File: docs/commerce/checkout/customize_checkout.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+The most important contract exposed by the package is the `CheckoutServiceInterface` interface. 
+It exposes a number of methods that you can call, for example, to load checkouts based 
+on checkout identifier or for a specific cart. 
+Other methods help you create, update, or delete checkouts. 
+
+For more information, see [Checkout API](checkout_api.md).
+
+## Add checkout stage
+
+By default [[= product_name =]] comes with a multi-step checkout process, which you can expand by adding steps.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:41]:
```suggestion
By default, [[= product_name =]] comes with a multi-step checkout process, which you can expand by adding steps.
```

File: docs/commerce/checkout/customize_checkout.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+For example, if you were creating a project for selling theater tickets, you could add a step 
+that allows users to select their seats.
+
+### Define workflow
+
+You can create workflow definitions in the `config/packages/ibexa.yaml` file. 
+Each workflow definition consists of a series of stages as well as a series of transitions between the stages.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:48]: I'm wondering if it's worth mentioning here that this workflow is Symfony workflow and different from the editorial workflow. Not sure.

File: docs/commerce/checkout/customize_checkout.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+By default [[= product_name =]] comes with a multi-step checkout process, which you can expand by adding steps.
+For example, if you were creating a project for selling theater tickets, you could add a step 
+that allows users to select their seats.
+
+### Define workflow
+
+You can create workflow definitions in the `config/packages/ibexa.yaml` file. 
+Each workflow definition consists of a series of stages as well as a series of transitions between the stages. 
+
+To create a new workflow, for example, `seat_selection_checkout`, modify the default workflow that comes with the storefront module, by adding a  `select_seat` stage.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:50]:
```suggestion
To create a new workflow, for example, `seat_selection_checkout`, modify the default workflow that comes with the storefront module, by adding a  `seat_selected` stage.
```
That's stage name, and `select_seat` is a transition, right?

File: docs/commerce/checkout/customize_checkout.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+### Create controller
+
+At this point you must add a controller that supports the newly added step.
+In this case, you want users to select seats in the audience.
+
+In the `src/Controller/Checkout/Step` folder, create a file that resembles the following example.
+
+The controller contains a Symfony form that collects user selections. 
+It can reuse fields and functions that come from the checkout component, for example, 
+after you check whether the form is valid, use the `advance` method to go to the next stage
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:72]:
```suggestion
after you check whether the form is valid, use the `AbstractStepController::advance` method to go to the next stage 
```

File: docs/commerce/checkout/customize_checkout.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+### Create Twig template
+
+You also need a Twig template to render the Symfony form.
+In `templates/themes/custom/storefront/checkout/step`, create a layout that uses JavaScript to translate clicking into a grid to a change in value:
+
+```html+twig
+[[= include_file('code_samples/front/shop/checkout/templates/themes/storefront/checkout/step/select_seat.html.twig') =]]
+```
+
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:87]: Can you add a screenshot to demonstrate how that looks?

File: docs/commerce/checkout/customize_checkout.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+### Configure repository
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:88]: I'd change the heading, you don't really configure the repository here, but rather select which workflow is used by the repository

File: docs/commerce/checkout/customize_checkout.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+At this point, you must inform the application about a repository that the workflow supports.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:90]:
```suggestion
At this point, you must inform the application that your repository will use the configured workflow.
```

File: docs/commerce/checkout/customize_checkout.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+You do it repository configuration, by replacing the `ibexa_checkout` configuration with one for `seat_selection_checkout`:
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:92]:
```suggestion
You do it in repository configuration, by replacing the `ibexa_checkout` configuration with one for `seat_selection_checkout`:
```

File: docs/commerce/checkout/customize_checkout.md [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    repositories:
+        default: 
+            checkout:
+                workflow: seat_selection_checkout
+```
+
+### Restart application
+
+You are now ready to see the results of your work.
+Shut down the application, clear browser cache, and restart the application.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:105]: Is shutting down the app necessary? And do you need to clear cache in  a more advanced way than refreshing the page?

File: docs/commerce/checkout/customize_checkout.md [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+You should be able to see a different checkout applied after you have added products to a cart.
+
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:107]: Or a screenshot here

File: docs/commerce/checkout/customize_checkout.md [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+- shipping method
+- billing / delivery address
+
+It could also collect any other information that you find necessary.
+
+Depending on your needs, the checkout process can be either complicated or very simple. 
+For example, if the website is selling airline tickets, you may need several additional steps 
+with passengers defining their special needs.
+On the other side of the spectrum would be a store that sells books with personal pickup, 
+where one-page checkout would be enough.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:24]: Maybe link to the two heading below, to make it clear that the page describes two separate procedures?

File: docs/commerce/checkout/customize_checkout.md [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+## Create a single-form checkout
+
+Another way of customizing the process would be to implement a single-form checkout.
+Such solution could work for certain industries, where simplicity is key.
+The single form's basic advantage is simplified navigation with less clicks to complete the transaction.
+
+### Define workflow
+
+To create a single-form checkout, define a workflow that has two stages, `initialized` and `completed`, and one transition, from `initialized` or `completed` to `completed`.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:116]: Can the transition go from completed to completed? It's not intuitive, so it requires an explanation.

File: docs/commerce/checkout/customize_checkout.md [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+
+``` yaml hl_lines="3 18 19"
+[[= include_file('code_samples/front/shop/checkout/config/packages/ibexa.yaml', 25, 27) =]] [[= include_file('code_samples/front/shop/checkout/config/packages/ibexa.yaml', 84, 103) =]] [[= include_file('code_samples/front/shop/checkout/config/packages/ibexa.yaml', 23, 24) =]]
+```
+
+### Create controller
+
+Add a controller in project code that is a regular Symfony controller, which reuses classes provided by the application.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:124]:
```suggestion
Add a regular Symfony controller, which reuses classes provided by the application.
```

File: docs/commerce/checkout/customize_checkout.md [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+Within the controller, create a form that contains all the necessary fields, such as the shipping and billing addresses, as well as shipping and billing methods.
+
+In the `src/Controller/Checkout` folder, create a file that resembles the following example:
+
+``` php
+[[= include_file('code_samples/front/shop/checkout/src/Controller/SinglePageCheckout.php') =]]
+```
+
+Again, after you handle the form and validate it, you advance the workflow by using the `advance` method provided in the base component.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:133]: I would not use "again" here, but repeat the explanation from above. This is a separate procedure than the one above and we can't assume readers will read the first one before starting with this one.

File: docs/commerce/checkout/customize_checkout.md [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+### Create Twig template
+
+Create a Twig template to render the Symfony form.
+In `templates/themes/custom/storefront/checkout`, create a layout that iterates through all the fields and renders them.
+
+```html+twig
+[[= include_file('code_samples/front/shop/checkout/templates/themes/storefront/checkout/checkout.html.twig') =]]
+```
+### Configure repository
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:143]: Same as above

File: docs/commerce/checkout/customize_checkout.md [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+ibexa:
+    repositories:
+        default: 
+            checkout:
+                workflow: single_page_checkout
+```
+
+### Restart application
+
+To see the results of your work, shut down the application, clear browser cache, and restart the application.
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:158]: Same as above

File: docs/commerce/checkout/customize_checkout.md [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+To create custom Address Field Type formats to be used in checkout, make the following changes in the project configuration files. 
+
+First, define custom format configuration keys for `billing_address_format` and `shipping_address_format`:
+
+``` yaml 
+ibexa:
+    repositories:
+        <repository_name>:
+            checkout:
+                #coming from Corporate Account, "billing" by default
```
Comment (DominikaK) [inline docs/commerce/checkout/customize_checkout.md:172]: Can you avoid mentioning corporate account here? It's confusing if the reader doesn't use it.

**Review by DominikaK — CHANGES_REQUESTED** [review]
General thought: maybe slip the page into two separate recipes for easier search and navigation?

File: code_samples/front/shop/checkout/templates/themes/storefront/checkout/checkout.html.twig [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+                    </div>
+                </fieldset>
+            {% endif %}
+        {% endfor %}
+
+        <button class="ibexa-store-btn ibexa-store-btn--primary">
+            Complete purchase
+        </button>
+    {{ form_end(form) }}
+{% endblock %}
```
Comment (konradoboza) [inline code_samples/front/shop/checkout/templates/themes/storefront/checkout/checkout.html.twig:38]:
```suggestion
{% endblock %}

```

File: code_samples/front/shop/checkout/templates/themes/storefront/checkout/step/select_seat.html.twig [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+                    const selection = [
+                        ...document.querySelectorAll(SELECTED_SEATS_SELECTOR)
+                    ].map((seat) => seat.dataset.seatId);
+
+                    input.value = selection.join(',');
+                });
+            });
+        })();
+    </script>
+{% endblock %}
```
Comment (konradoboza) [inline code_samples/front/shop/checkout/templates/themes/storefront/checkout/step/select_seat.html.twig:73]:
```suggestion
{% endblock %}

```

File: code_samples/front/shop/checkout/templates/themes/storefront/layout.html.twig [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+{% endblock %}
+
+{% block main_menu %}
+    {% set main_menu_alias = ibexa_storefront_get_main_menu_alias() %}
+    {% set main_menu_options = ibexa_storefront_get_main_menu_options() %}
+
+    <div class="ibexa-store-main-menu-items">
+        {{ knp_menu_render(main_menu_alias, main_menu_options) }}
+    </div>
+{% endblock %}
```
Comment (konradoboza) [inline code_samples/front/shop/checkout/templates/themes/storefront/layout.html.twig:20]:
```suggestion
{% endblock %}

```

File: docs/commerce/checkout/customize_checkout.md (hunk 3 above)
Comment (konradoboza) [reply docs/commerce/checkout/customize_checkout.md:48]: Definitely, Ibexa Workflow operates only in the context of content objects which isn't the case in the checkout.

File: docs/commerce/checkout/customize_checkout.md (hunk 4 above)
Comment (konradoboza) [reply docs/commerce/checkout/customize_checkout.md:50]: Actually it's referred to as "step". in the code. Transition is a Symfony workflow term but both (step, transition) can be used.

File: docs/commerce/checkout/customize_checkout.md (hunk 10 above)
Comment (konradoboza) [reply docs/commerce/checkout/customize_checkout.md:105]: Probably even clearing the cache is not mandatory, but I would leave it.

File: docs/commerce/checkout/customize_checkout.md (hunk 3 above)
Comment (dabrt) [reply docs/commerce/checkout/customize_checkout.md:48]: @DominikaK I am mentioning it in line 28, do you think it is necessary to repeat this infornation here?

File: docs/commerce/checkout/customize_checkout.md [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+Add a regular Symfony controller in project code, which reuses classes provided by the application.
+Within the controller, create a form that contains all the necessary fields, such as the shipping and billing addresses, as well as shipping and billing methods.
+
+In the `src/Controller/Checkout` folder, create a file that resembles the following example:
+
+``` php
+[[= include_file('code_samples/front/shop/checkout/src/Controller/SinglePageCheckout.php') =]]
+```
+
+IThe controller can reuse fields and functions that come from the checkout component, for example,
```
Comment (konradoboza) [inline docs/commerce/checkout/customize_checkout.md:134]:
```suggestion
The controller can reuse fields and functions that come from the checkout component, for example, 
```

**Review by konradoboza — APPROVED** [review]

**Comment by julitafalcondusza** [comment]
Please add interactive link to JIRA ticket in the table of PR.

**Review by julitafalcondusza — APPROVED** [review]

**Pull request merged**